project/iar1000/README.md

```diff
@@ -0,0 +1,98 @@
+
+#### Point Event
+
+The sweep line encounters target point $t_k$. The horizontal line $h_k$ going through $t_k$ is going to intersect with a parabola $p_k \in P$ to the left of the sweep line.
+
+A new parabola is introduced to the beach line at the intersection of $h_k$ and $p_k$. It is infitesimally small in the beginning and starts to grow as the sweep line progresses, tracking the equidistance level curve for the newly incorporated point $t_k$.
+
+As the sweep line progresses only in one direction, we know the order of the point events in advance. Sorting the target points in advance allows us an efficient lookup for the next point event.
+
+#### Circle Event
+
+Circle events are created on the beach line by three neighbouring target points. Each of the three points has an arc associated with it, all part of the beach line, which indicates equidistance from the closest target point and the sweep line. While all of the target points grow their cell, their respective arcs on the beach line grow with them, until one of them is "squashed" to length zero by the other two. This is going to happen at point $v_i$, which has the same distance to all three neighbours. A Voronoi vertex is created at this location and the parabola with empty length is removed, because we know that in the final diagram, a border is starting at that point.
+
+### Pseudo Code
+
+We keep track of the events in a event queue $Q$, handling them one after the other.
+
+```
+insert all Point Events to Q, sorted by x-coordinate
+while Q is not empty:
+    pop event E
+    if E is a Point Event:
+        insert new arc to the beach line
+        check for new Circle Events
+    else:
+        create a new vertex in the diagram D
+        remove the shrunk arc from the beach line
+        delete invalidated events
+        check and add new Circle Events
+```
+
+### Interfaces
+
+The software will have one central datastructure, namely the ``VoronoiDiagram``. The Voronoi diagram $V$ is implemented as a [Doubly Connected Edge List (DCEL)](https://en.wikipedia.org/wiki/Doubly_connected_edge_list), which stores the planar graph as a set of (half-)edges, vertices and cells. It is special in that each edge is represented as two half-edges, each associated to one of the cells forming the edge.
+As input the software takes a ``.csv`` or ``.json`` file with target points, from where it computes a valid Voronoi diagram. The output is written to a ``.json`` file describing all vertices and edges of the diagram. The bounding box used is defined by the min/max coordinates of all points.
+
+## API Description
+
+Will follow when coding has started
+
+## Engineering Infrastructure
+
+* Build: `make`
+* Version control: `git`
+* Testing infrastructure: `pytest`
+* Documentation: `README.md` and inline
+
+## Schedule
+
+Week 1 (10/31): C++ reading input and writing output
```

Comment:
You could find that to read input from Python and convert to C++ is an easier way to do it.

Reply:
Is the end goal to use it as python module? 
I interpreted it by now as having to make a stand-alone c++ program which can be used with python aswell, but if I can reduce it to making it a python module with a c++ engine it would be great. Then I could do the reading and writing with python

Reply:
Yes, but if you want to do otherwise it's negotiable.  I very much welcome technical arguments.

As you will see later in the course, we make a lot of use of Python interpreter.  Scientists and engineers love interpreters, and Python is one of the best.